### PR TITLE
support cobo vault

### PR DESCRIPTION
**Pull request by soralit**
This PR adds Cobo Vault interaction for Caravan-CoboVault integration usage

### CODE REVIEW

**Review by waldenraines — CHANGES_REQUESTED** [review]
When attempting to import the multisig wallet into the cobo I get:

```
This QR Code is not multisig wallet, can't import
```

This prevents me from being able to test signing.  Unsure if the problem is in caravan or here.

File: package.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
   "bin": {},
   "dependencies": {
     "@babel/polyfill": "^7.7.0",
+    "@cvbb/bc-ur": "^0.2.8",
```
Comment (humanumbrella) [inline package.json:44]: this went up to `0.2.9`

File: src/index.js [hunk 2]
```diff
@@ -302,6 +303,14 @@ export function SignMultisigTransaction({keystore, network, inputs, outputs, bip
         bip32Paths,
         psbt,
       });
+    case COBOVAULT:
```
Comment (humanumbrella) [inline src/index.js:306]: Let's add `COBOVAULT` to "Indirect Keystores" above

File: src/index.js [hunk 3]
```diff
@@ -28,6 +28,7 @@ import {
   ColdcardSignMultisigTransaction,
   ColdcardMultisigWalletConfig,
 } from './coldcard';
+import {COBOVAULT, CoboVaultSignMultisigTransaction} from "./cobovault";
```
Comment (humanumbrella) [inline src/index.js:31]: Will want to add support here for `CoboVaultExportExtendedPublicKey`, no? Right now only exposing `CoboVaultSignMultisigTransaction` via this index file, though it looks like extendedpublickey is implemented in `cobovault.js`

File: src/coldcard.js [hunk 4]
```diff
@@ -29,33 +29,27 @@ import {
   INFO,
   ERROR,
 } from "./interaction";
-import {
-  P2SH,
-  P2SH_P2WSH,
-  P2WSH,
-} from 'unchained-bitcoin';
+import { P2SH, P2SH_P2WSH, P2WSH } from "unchained-bitcoin";
```
Comment (humanumbrella) [inline src/coldcard.js:32]: don't think this file needs to be touched

File: src/coldcard.fixtures.js [hunk 5]
```diff
@@ -1,77 +1,97 @@
-import {COLDCARD_WALLET_CONFIG_VERSION} from './coldcard';
+import { COLDCARD_WALLET_CONFIG_VERSION } from "./coldcard";
```
Comment (humanumbrella) [inline src/coldcard.fixtures.js:1]: same (no need to touch coldcard stuff)

File: src/cobovault.js [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+ *     result: string | object
+ *     type: 'text' | 'ur' | 'json'
+ *     current: number,
+ *     total: number,
+ *     workloads: array<string>
+ * }
+ */
+function smartParseQR(qrs) {
+    const enhancedQRS = qrs.filter(onlyUniq);
+    const first = qrs[0];
```
Comment (humanumbrella) [inline src/cobovault.js:32]: Would it not be better to use `enhancedQRs[0]` here instead of `qrs[0]` ?

File: src/cobovault.js [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        if (first.startsWith("UR")) {
+            //UR
+            return parseUR(enhancedQRS)
+        }
+        else {
+            //raw text
+            return {
+                success: true,
+                result: first,
+                type: 'text',
```
Comment (humanumbrella) [inline src/cobovault.js:47]: 'text' => "text"

File: src/cobovault.js [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+ */
+export class CoboVaultDisplayer extends CoboVaultInteraction {
+
+    constructor() {
+        super();
+        this.displayer = true;
+    }
+
+    encodeUR(data) {
+        return encodeUR(data, 800);
```
Comment (humanumbrella) [inline src/cobovault.js:161]: maybe a comment about this hardcoded value?

File: src/cobovault.js [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+ * // "xpub..."
+ * console.log(bip32Path);
+ * // "m/45'/0'/0'"
+ */
+
+export class CoboVaultExportExtendedPublicKey extends CoboVaultReader {
+    messages() {
+        const messages = super.messages();
+        messages.push(commandMessage({
+            instructions: "①Please open Cobo Vault, go to Menu > Multisig Wallet > More > Show/Export  XPUB. ②Click the camera icon above and scan the QR Code of XPUB displays on Cobo Vault",
```
Comment (humanumbrella) [inline src/cobovault.js:184]: this might be better as multiple messages instead of one with multiple instructions inside

**Comment by waldenraines** [comment]
I think this PR needs to be updated with a new run of `npm update`.  I get the following differences in `package-lock.json` when installing fresh:

```
[walden:~/code/unchained-wallets] 3e1ddc2(+10/-11)+ ± git diff
diff --git a/package-lock.json b/package-lock.json
index c391ea9..b30d07b 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -2073,15 +2073,14 @@
       "integrity": "sha512-i3X26uKJOkDTAalYAp0Er+qGMDhrbbh2o93/xiPyAN2s25KrClSpe3VXo/7mNJoqA5qfko8rLS2l3RWZgYmjKQ=="
     },
     "bip32": {
-      "version": "2.0.6",
-      "resolved": "https://registry.npmjs.org/bip32/-/bip32-2.0.6.tgz",
-      "integrity": "sha512-HpV5OMLLGTjSVblmrtYRfFFKuQB+GArM0+XP8HGWfJ5vxYBqo+DesvJwOdC2WJ3bCkZShGf0QIfoIpeomVzVdA==",
+      "version": "1.0.4",
+      "resolved": "https://registry.npmjs.org/bip32/-/bip32-1.0.4.tgz",
+      "integrity": "sha512-8T21eLWylZETolyqCPgia+MNp+kY37zFr7PTFDTPObHeNi9JlfG4qGIh8WzerIJidtwoK+NsWq2I5i66YfHoIw==",
       "requires": {
-        "@types/node": "10.12.18",
         "bs58check": "^2.1.1",
         "create-hash": "^1.2.0",
         "create-hmac": "^1.1.7",
-        "tiny-secp256k1": "^1.1.3",
+        "tiny-secp256k1": "^1.0.0",
         "typeforce": "^1.11.5",
         "wif": "^2.0.6"
       }
@@ -2115,13 +2114,12 @@
       "integrity": "sha512-pef6gxZFztEhaE9RY9HmWVmiIHqCb2OyS4HPKkpc6CIiiOa3Qmuoylxc5P2EkU3w+5eTSifI9SEZC88idAIGow=="
     },
     "bitcoinjs-lib": {
-      "version": "5.2.0",
-      "resolved": "https://registry.npmjs.org/bitcoinjs-lib/-/bitcoinjs-lib-5.2.0.tgz",
-      "integrity": "sha512-5DcLxGUDejgNBYcieMIUfjORtUeNWl828VWLHJGVKZCb4zIS1oOySTUr0LGmcqJBQgTBz3bGbRQla4FgrdQEIQ==",
+      "version": "4.0.5",
+      "resolved": "https://registry.npmjs.org/bitcoinjs-lib/-/bitcoinjs-lib-4.0.5.tgz",
+      "integrity": "sha512-gYs7K2hiY4Xb96J8AIF+Rx+hqbwjVlp5Zt6L6AnHOdzfe/2tODdmDxsEytnaxVCdhOUg0JnsGpl+KowBpGLxtA==",
       "requires": {
         "bech32": "^1.1.2",
-        "bip174": "^2.0.1",
-        "bip32": "^2.0.4",
+        "bip32": "^1.0.4",
         "bip66": "^1.1.0",
         "bitcoin-ops": "^1.4.0",
         "bs58check": "^2.0.0",
@@ -2130,7 +2128,8 @@
         "merkle-lib": "^2.0.10",
         "pushdata-bitcoin": "^1.0.1",
         "randombytes": "^2.0.1",
-        "tiny-secp256k1": "^1.1.1",
+        "safe-buffer": "^5.1.1",
+        "tiny-secp256k1": "^1.0.0",
         "typeforce": "^1.11.3",
         "varuint-bitcoin": "^1.0.4",
         "wif": "^2.0.1"
```

File: src/cobovault.js [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+export class CoboVaultSignMultisigTransaction extends CoboVaultReader {
+
+    /**
+     *
+     * @param {object} options - options argument
+     * @param {array<object>} options.inputs - inputs for the transaction
+     * @param {array<object>} options.outputs - outputs for the transaction
+     * @param {array<string>} options.bip32Paths - BIP32 paths
+     */
+    constructor({network, inputs, outputs, bip32Paths, psbt}) {
```
Comment (humanumbrella) [inline src/cobovault.js:258]: if non-P2SH inputs or outputs are not supported by CoboVault - check it here and fail appropriately.

**Review by humanumbrella — CHANGES_REQUESTED** [review]
Sorry for the delay - I have posted some comments and questions.

**Comment by humanumbrella** [comment]
Can you please add unit tests ?
Screen Shot 2021-02-02 at 1 45 30 PM

File: src/cobovault.js [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+        }
+    }
+}
+
+/**
+ * Returns signature request data to display in a QR code for CoboVault
+ * and reads the signature data passed back by CoboVault in another QR
+ * code.
+ *
+ * NOTE: Transactions with inputs & outputs to non-P2SH addresses are not supported by CoboVault.
```
Comment (humanumbrella) [inline src/cobovault.js:232]: I do not think this ^ is true, maybe came over from the `hermit.js` file.

File: src/coldcard.js (hunk 4 above)
Comment (soralit) [reply src/coldcard.js:32]: Seems introducing it by rebasing master.

File: src/cobovault.js (hunk 6 above)
Comment (soralit) [reply src/cobovault.js:32]: 👍

File: src/cobovault.js (hunk 7 above)
Comment (soralit) [reply src/cobovault.js:47]: Maybe a lint rule would be better?

**Comment by soralit** [comment]
Hi @humanumbrella , I have fixed code issues and add some unit tests, would you like to have a review when you have time?

File: package.json [hunk 12]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "unchained-wallets",
-  "version": "0.1.14",
+  "version": "0.1.15",
```
[the remark this answers is not in the document]
Comment (humanumbrella) [reply package.json]: ignore, sorry

**Review by bucko13 — CHANGES_REQUESTED** [review]
First, sorry for such a long delay in getting a proper review to you all. Most of our comments are around things that we hadn't properly documented in what the expected API for unchained-wallets classes should look like. If you have any questions about those, feel free to ask.

Also (another thing not documented. Sorry!), please leave the lock file out of the commit. We want it in version control for now, but will commit it during the release process. Same comment for the version upgrade in the package.json file.

**Comment by waldenraines** [comment]
Closing since https://github.com/unchained-capital/caravan/pull/185 was closed and I'm assuming this will also be redone.   Please reopen if not.

**Pull request closed without merging**